### PR TITLE
✅[outreachy] Registry : Add to favorites #5447

### LINKED ISSUE #5447: [outreachy] Registry: Add to favorites
Outreachy task for @mailsg

Based on your user experience report, please provide the following suggested improvement via a PR:

> For quick engagement with each registry item, consider adding a set of quick action icons next to each entry. These could include a star for “Add to Favorites”, allowing users to save items for later [...]

Please only add a quick action for "add to favorites". You can store the favorites in a cookie or local storage.

**Note**: This is an assigned issue for a participant that has completed the user experience report, if you are an outreachy participant as well, find instructions here: #5290

### PR DESCRIPTION
**Pull request by mailsg**
## Fixes
- Fixes #5447 by @svrnm

## Description
- This PR introduces a "Add to Favorites" feature, allowing users to quickly mark registry entries as favorites. Favorites are stored locally in local storage, ensuring persistence across sessions.
    - Implemented a button for adding/removing items from favorites.
    - Favorites are stored in local storage for better user experience.
    - Integrated the functionality using a `favorite-btn` that updates dynamically.

## Technical details
- Partial for JavaScript: Moved JavaScript functionality to a separate `js/addToFavorites.js file`, loaded via a partial in the layout for better maintainability and performance.

- Local Storage Logic: The local storage checks if an item is already marked as a favorite, and updates the button's appearance accordingly. This avoids the need for server-side calls, keeping it fast and client-side focused.

## Screenshots

SS1

SS2

SS3

### CODE REVIEW

**Comment by svrnm** [comment]
@mailsg: I see that I can add favorites, have you planned any way that I can _use_ the favorites, e.g. only see the favorites or any other way?

**Comment by svrnm** [comment]
@mailsg please make sure that you register with the outreachy platform after that I will provide a review

**Comment by svrnm** [comment]
thanks for registering your contributions in the outreachy platform. Following up with my previous feedback:

I see that I can add favorites, have you planned any way that I can use the favorites, e.g. only see the favorites or any other way?

**Comment by mailsg** [comment]
> thanks for registering your contributions in the outreachy platform. Following up with my previous feedback:
> 
> I see that I can add favorites, have you planned any way that I can use the favorites, e.g. only see the favorites or any other way?

Hi @svrnm,

In addition to providing users with the ability to add and remove items from their favorites, there are several practical ways to make this feature more useful. These can be the planned usage of the `Add to Favorites` feature :

1. **Dedicated "Favorites" Page**
   - Users will be able to view all of their favorite marked items in one place by navigating to a dedicated "Favorites" page. This page will aggregate all marked items for easy access.
   - This allows users to quickly locate their preferred resources without having to search through the entire registry each time.

2. **Filter by Favorites**
   - An additional filter could be added to the registry that allows users to view only their favorite marked items while browsing the main catalog. This would integrate the favorites functionality directly into the registry interface, making it more user-friendly and efficient for users looking to engage with their curated resources. 

3. **Persistence Across Sessions**
   - Since the favorites are stored in local storage or session cookies, users can access their list of favorites even after logging out or closing the browser. This ensures that the favorites list persists across sessions, improving the user experience by providing continuity. 

4. **Improved Personalization**
   - With a user's favorites data available, the site could later introduce personalized content suggestions based on their favorites, such as related tools, newly added components, or relevant updates. This could lead to better user satisfaction and engagement without requiring significant changes to the initial feature implementation.

5. **Future Expansions**
   - In the future, favorites could serve as the foundation for more advanced features, such as exporting favorite lists, sharing with others, or receiving notifications when an item in their favorites is updated. These expansions would make the feature even more valuable to users while aligning with the overall goal of improving user experience.

To conclude, the `Add to Favorites` feature is designed to make the registry more user-friendly and personalized. By giving users a dedicated page for their favorites, a filter to view only their selected items, and ensuring everything is saved across sessions, we are making it easier for users to engage with the platform in a way that works for them. This will not only make the registry more useful on a day-to-day basis, but it also sets the stage for future improvements like personalized recommendations or sharing favorites with others. Overall, I think this can be a step towards making the site more interactive and valuable for the community.

Thanks.

**Comment by svrnm** [comment]
Please implement "Filter by Favorites". Thanks.

**Comment by mailsg** [comment]
> Please implement "Filter by Favorites". Thanks.

### Description
- The JavaScript functionality enhances user interaction with the registry by allowing users to filter their view to display only their favorite entries. This feature simplifies navigation by letting users focus on items they are most interested in.

### Technical details
- The implementation consists of event listeners that manage user interactions with the `Show Favorites` button. When a user selects to view only their favorites, the system checks against local storage to display the appropriate items and toggling `Show Favorites` to `Show All`.
-  When the user selects `Show All`, they can view all the registry items.


SS-1



SS-2


Thanks.

File: .htmltest.yml [hunk 1]
```diff
@@ -8,11 +8,11 @@ CheckMailto: false
 TestFilesConcurrently: true
 IgnoreDirs:
   # DO NOT EDIT! IgnoreDirs list is auto-generated from markdown file front matter.
+  - ^blog/(\d+/)?page/\d+
   # TODO drop next line after https://github.com/open-telemetry/opentelemetry.io/issues/5423 is fixed for ja pages:
   - ^ja/docs/concepts/instrumentation/libraries/
   # TODO drop next line after https://github.com/open-telemetry/opentelemetry.io/issues/5423 is fixed for pt pages:
   - ^pt/docs/concepts/instrumentation/libraries/
-  - ^blog/(\d+/)?page/\d+
```
Comment (svrnm) [inline .htmltest.yml:15]: please revert this change

File: assets/js/addToFavorites.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        favorites = favorites.filter((id) => id !== registryId);
+        localStorage.setItem('favorites', JSON.stringify(favorites));
+        button.innerHTML =
+          '<i class="fa-regular fa-star"></i> Add to favorites';
+        button.classList.add('btn-outline-primary');
+        button.classList.remove('btn-success');
+      } else {
+        favorites.push(registryId);
+        localStorage.setItem('favorites', JSON.stringify(favorites));
+        button.innerHTML = '<i class="fa-solid fa-star"></i>';
```
Comment (svrnm) [inline assets/js/addToFavorites.js:24]: Remove from favorites to make it clear that it is a favorite and can be removed by clicking on the star once again

**Comment by mailsg** [comment]
> overall lgtm, a few comments

Hi @svrnm,

I have made the suggested changes. 

SS-1


Thanks.

**Comment by svrnm** [comment]
I consider this as done! After the outreachy application phase we will take another look and see if we can merge this PR. Thank you!

**Comment by svrnm** [comment]
I will close this for now, since I raised a follow up discussion here: https://github.com/open-telemetry/opentelemetry.io/issues/5759
